### PR TITLE
Forward port java annotation handling

### PR DESCRIPTION
**Pull request by som-snytt**
Further improve

https://github.com/scala/scala/pull/9981

by forward porting

https://github.com/scala/scala/pull/10016

and then get carried away a bit with tweaks.

### CODE REVIEW

File: src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala [hunk 1]
```diff
@@ -616,26 +612,19 @@ abstract class ClassfileParser(reader: ReusableInstance[ReusableDataReader]) {
   def parseMethod(): Unit = {
     val jflags = readMethodFlags()
     val sflags = jflags.toScalaFlags
-    if (jflags.isPrivate) {
-      val isConstructor = pool.getName(u2()).value == "<init>" // opt avoid interning a Name for private methods we're about to discard
-      if (isConstructor)
-        sawPrivateConstructor = true
-      in.skip(2); skipAttributes()
+    if (jflags.isPrivate || ((sflags & PRIVATE) != 0L)) {
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala]: Not sure if these tests are meaningfully distinct.

Comment (lrytz) [reply src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala]: They are the same 👍

Comment (som-snytt) [reply src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala]: Thanks LKAS @lrytz

**Review by lrytz — APPROVED** [review]

File: src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala [hunk 2]
```diff
@@ -865,7 +854,7 @@ abstract class ClassfileParser(reader: ReusableInstance[ReusableDataReader]) {
         case tpnme.RuntimeAnnotationATTR =>
             val numAnnots = u2()
           val annots = new ListBuffer[AnnotationInfo]
-            for (n <- 0 until numAnnots; annot <- parseAnnotation(u2()))
+          for (n <- 0 until numAnnots; annot <- parseAnnotation(u2()))
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala:857]: Distracted by indent, forgot to use fancy `times`.

File: src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala [hunk 3]
```diff
@@ -1502,3 +1492,8 @@ abstract class ClassfileParser(reader: ReusableInstance[ReusableDataReader]) {
   protected def getScope(flags: JavaAccFlags): Scope =
     if (flags.isStatic) staticScope else instanceScope
 }
+object ClassfileParser {
+  private implicit class GoodTimes(val n: Int) extends AnyVal {
+    def times(body: => Unit) = (1 to n).foreach(_ => body)
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala:1497]: Could have gone `while` here.

**Pull request merged**